romancal/source_catalog/source_catalog_step.py

```diff
@@ -0,0 +1,123 @@
+        suffix = string(default='cat')        # Default suffix for output files
+    """
+
+    def _get_reffile_paths(self, model):
+        filepaths = []
+        for reffile_type in self.reference_file_types:
+            try:
+                filepath = self.get_reference_file(model, reffile_type)
+                self.log.info(
+                    f"Using {reffile_type.upper()} reference file: " f"{filepath}"
+                )
+            except CrdsLookupError as err:
+                msg = f"{err} Source catalog will not be created."
+                self.log.warning(msg)
+                return None
+
+            filepaths.append(filepath)
+        return filepaths
+
+    def process(self, input_model):
+        with datamodels.open(input_model) as model:
+            reffile_paths = self._get_reffile_paths(model)
+            aperture_ee = (self.aperture_ee1, self.aperture_ee2, self.aperture_ee3)
+
+            try:
+                refdata = ReferenceData(model, reffile_paths, aperture_ee)
+                aperture_params = refdata.aperture_params
+
+                # TODO (@bmorris3): replace or remove
+                abvega_offset = refdata.abvega_offset
+            except RuntimeError as err:
+                msg = f"{err} Source catalog will not be created."
+                self.log.warning(msg)
+                return None
+
+            coverage_mask = np.isnan(model.err)  # | (model.wht == 0)
+            bkg = RomanBackground(
+                model.data, box_size=self.bkg_boxsize, coverage_mask=coverage_mask
+            )
+            model.data -= bkg.background
+
+            threshold = self.snr_threshold * bkg.background_rms
+            finder = RomanSourceFinder(threshold, self.npixels, deblend=self.deblend)
+
+            convolved_data = convolve_data(
+                model.data, self.kernel_fwhm, mask=coverage_mask
+            )
+            segment_img = finder(convolved_data, mask=coverage_mask)
+            if segment_img is None:
+                return None
+
+            ci_star_thresholds = (self.ci1_star_threshold, self.ci2_star_threshold)
+            catobj = RomanSourceCatalog(
+                model,
+                segment_img,
+                convolved_data,
+                self.kernel_fwhm,
+                aperture_params,
+                abvega_offset,
+                ci_star_thresholds,
+            )
+            catalog = catobj.catalog
+
+            # add back background to data so input model is unchanged
+            model.data += bkg.background
+
+            if self.save_results:
+                cat_filepath = self.make_output_path(ext=".ecsv")
+                catalog.write(cat_filepath, format="ascii.ecsv", overwrite=True)
+                model.meta.source_catalog = os.path.basename(cat_filepath)
+                self.log.info(f"Wrote source catalog: {cat_filepath}")
+
+                segm_model = datamodels.SegmentationMapModel(segment_img.data)
```

Comment:
cc @nden, @schlafly: We haven't yet discussed if I should implement the Segmentation Map Model as part of this effort. Should I?

Reply:
Added by @nden